File: Sources/BibleVerseReplacer/ReferenceParser.swift
import Foundation

final class ReferenceParser {
    func parseSelection(_ rawSelection: String) throws -> ParsedReference {
        let normalized = normalizeSelection(rawSelection)
        guard !normalized.isEmpty else {
            throw ReferenceParseError.emptySelection
        }

        let chunks = normalized
            .split(separator: ",")
            .map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        guard !chunks.isEmpty else {
            throw ReferenceParseError.unrecognizedReference
        }

        var passages: [PassageReference] = []
        var currentBook: BibleBook?
        var currentChapter: Int?

        for chunk in chunks {
            let parsed = try parseChunk(chunk, currentBook: currentBook, currentChapter: currentChapter)
            passages.append(parsed.passage)
            currentBook = parsed.passage.book
            currentChapter = parsed.contextChapter
        }

        return ParsedReference(passages: passages)
    }

    func parse(_ rawSelection: String) throws -> VerseReference {
        let parsed = try parseSelection(rawSelection)
        guard parsed.passages.count == 1, let passage = parsed.passages.first,
              passage.startChapter == passage.endChapter,
              let startVerse = passage.startVerse,
              let endVerse = passage.endVerse else {
            throw ReferenceParseError.unrecognizedReference
        }

        return VerseReference(book: passage.book, chapter: passage.startChapter, startVerse: startVerse, endVerse: endVerse)
    }

    private func normalizeSelection(_ raw: String) -> String {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        text = text.replacingOccurrences(of: "\n", with: " ")
        text = text.replacingOccurrences(of: "\t", with: " ")
        text = text.replacingOccurrences(of: "\u{3000}", with: " ")
        text = text.replacingOccurrences(of: "：", with: ":")
        text = text.replacingOccurrences(of: "﹕", with: ":")
        text = text.replacingOccurrences(of: "“", with: "")
        text = text.replacingOccurrences(of: "”", with: "")
        text = text.replacingOccurrences(of: "\"", with: "")
        text = text.replacingOccurrences(of: "'", with: "")
        text = convertFullWidthDigits(text)
        text = normalizeRanges(text)
        text = normalizeSeparators(text)

        while text.contains("  ") {
            text = text.replacingOccurrences(of: "  ", with: " ")
        }
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func normalizeRanges(_ raw: String) -> String {
        var text = raw
        let replacements = [
            "……", "...", "——", "--", "－", "–", "—", "―", "﹣", "～", "~", "^", "到", "至"
        ]
        for token in replacements {
            text = text.replacingOccurrences(of: token, with: "-")
        }

        if let regex = try? NSRegularExpression(pattern: #"(?i)(\d)\s*to\s*(\d)"#) {
            let range = NSRange(text.startIndex..<text.endIndex, in: text)
            text = regex.stringByReplacingMatches(in: text, range: range, withTemplate: "$1-$2")
        }
        return text
    }

    private func normalizeSeparators(_ raw: String) -> String {
        var text = raw
        for token in ["，", "、", "；", ";", "｜", "|", "\\"] {
            text = text.replacingOccurrences(of: token, with: ",")
        }
        return text
    }

    private func convertFullWidthDigits(_ text: String) -> String {
        var result = ""
        for scalar in text.unicodeScalars {
            let value = scalar.value
            if value >= 0xFF10 && value <= 0xFF19 {
                result.unicodeScalars.append(UnicodeScalar(value - 0xFF10 + 0x30)!)
            } else {
                result.unicodeScalars.append(scalar)
            }
        }
        return result
    }

    private func parseChunk(
        _ rawChunk: String,
        currentBook: BibleBook?,
        currentChapter: Int?
    ) throws -> (passage: PassageReference, contextChapter: Int) {
        let compact = rawChunk
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: " ", with: "")

        guard !compact.isEmpty else {
            throw ReferenceParseError.unrecognizedReference
        }

        var book = currentBook
        var body = compact
        if let match = BibleBookCatalog.bookAtStart(of: compact) {
            book = match.book
            body = match.remaining
        }

        guard let resolvedBook = book else {
            throw ReferenceParseError.unknownBook(rawChunk)
        }

        if let passage = try parseChapterStyle(body, book: resolvedBook, currentChapter: currentChapter) {
            return (passage, passage.endChapter)
        }

        if let passage = try parseColonStyle(body, book: resolvedBook) {
            return (passage, passage.endChapter)
        }

        if let passage = try parseInheritedVerseStyle(body, book: resolvedBook, currentChapter: currentChapter) {
            return (passage, passage.endChapter)
        }

        throw ReferenceParseError.unrecognizedReference
    }

    private func parseChapterStyle(_ body: String, book: BibleBook, currentChapter: Int?) throws -> PassageReference? {
        guard body.contains("章") || body.contains("节") else {
            return nil
        }

        if let match = firstMatch(pattern: #"^第?(\d+)章(?:第?(\d+)(?:-(\d+))?节?)?$"#, in: body) {
            let chapter = intGroup(match, 1, in: body)
            let startVerse = optionalIntGroup(match, 2, in: body)
            let endVerse = optionalIntGroup(match, 3, in: body) ?? startVerse
            if let startVerse, let endVerse, startVerse > endVerse {
                throw ReferenceParseError.invalidRange
            }
            return PassageReference(book: book, startChapter: chapter, startVerse: startVerse, endChapter: chapter, endVerse: endVerse)
        }

        if let match = firstMatch(pattern: #"^第?(\d+)(?:-(\d+))?节$"#, in: body), let currentChapter {
            let startVerse = intGroup(match, 1, in: body)
            let endVerse = optionalIntGroup(match, 2, in: body) ?? startVerse
            if startVerse > endVerse {
                throw ReferenceParseError.invalidRange
            }
            return PassageReference(book: book, startChapter: currentChapter, startVerse: startVerse, endChapter: currentChapter, endVerse: endVerse)
        }

        return nil
    }

    private func parseColonStyle(_ body: String, book: BibleBook) throws -> PassageReference? {
        if let match = firstMatch(pattern: #"^(\d+):(\d+)-(\d+):(\d+)$"#, in: body) {
            let startChapter = intGroup(match, 1, in: body)
            let startVerse = intGroup(match, 2, in: body)
            let endChapter = intGroup(match, 3, in: body)
            let endVerse = intGroup(match, 4, in: body)
            if startChapter > endChapter || (startChapter == endChapter && startVerse > endVerse) {
                throw ReferenceParseError.invalidRange
            }
            return PassageReference(book: book, startChapter: startChapter, startVerse: startVerse, endChapter: endChapter, endVerse: endVerse)
        }

        if let match = firstMatch(pattern: #"^(\d+):(\d+)-(\d+)$"#, in: body) {
            let chapter = intGroup(match, 1, in: body)
            let startVerse = intGroup(match, 2, in: body)
            let endVerse = intGroup(match, 3, in: body)
            if startVerse > endVerse {
                throw ReferenceParseError.invalidRange
            }
            return PassageReference(book: book, startChapter: chapter, startVerse: startVerse, endChapter: chapter, endVerse: endVerse)
        }

        if let match = firstMatch(pattern: #"^(\d+):(\d+)$"#, in: body) {
            let chapter = intGroup(match, 1, in: body)
            let verse = intGroup(match, 2, in: body)
            return PassageReference(book: book, startChapter: chapter, startVerse: verse, endChapter: chapter, endVerse: verse)
        }

        return nil
    }

    private func parseInheritedVerseStyle(_ body: String, book: BibleBook, currentChapter: Int?) throws -> PassageReference? {
        if let match = firstMatch(pattern: #"^(\d+)-(\d+)$"#, in: body), let currentChapter {
            let startVerse = intGroup(match, 1, in: body)
            let endVerse = intGroup(match, 2, in: body)
            if startVerse > endVerse {
                throw ReferenceParseError.invalidRange
            }
            return PassageReference(book: book, startChapter: currentChapter, startVerse: startVerse, endChapter: currentChapter, endVerse: endVerse)
        }

        if let match = firstMatch(pattern: #"^(\d+)$"#, in: body) {
            let number = intGroup(match, 1, in: body)
            if let currentChapter {
                return PassageReference(book: book, startChapter: currentChapter, startVerse: number, endChapter: currentChapter, endVerse: number)
            }
            return PassageReference(book: book, startChapter: number, startVerse: nil, endChapter: number, endVerse: nil)
        }

        return nil
    }

    private func firstMatch(pattern: String, in text: String) -> NSTextCheckingResult? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else {
            return nil
        }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        return regex.firstMatch(in: text, range: range)
    }

    private func intGroup(_ match: NSTextCheckingResult, _ index: Int, in text: String) -> Int {
        Int(group(match, index, in: text)) ?? 0
    }

    private func optionalIntGroup(_ match: NSTextCheckingResult, _ index: Int, in text: String) -> Int? {
        let value = group(match, index, in: text)
        return value.isEmpty ? nil : Int(value)
    }

    private func group(_ match: NSTextCheckingResult, _ index: Int, in text: String) -> String {
        guard index < match.numberOfRanges else {
            return ""
        }
        let range = match.range(at: index)
        guard range.location != NSNotFound, let swiftRange = Range(range, in: text) else {
            return ""
        }
        return String(text[swiftRange])
    }
}

File: Windows/BibleVerseReplacer.Windows/ReferenceParser.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace BibleVerseReplacer.Windows
{
    internal sealed class ReferenceParser
    {
        private static readonly Regex CrossChapterRegex = new Regex(@"^(\d+):(\d+)-(\d+):(\d+)$", RegexOptions.Compiled);
        private static readonly Regex SameChapterRangeRegex = new Regex(@"^(\d+):(\d+)-(\d+)$", RegexOptions.Compiled);
        private static readonly Regex SingleVerseRegex = new Regex(@"^(\d+):(\d+)$", RegexOptions.Compiled);
        private static readonly Regex WholeChapterRegex = new Regex(@"^第?(\d+)章(?:第?(\d+)(?:-(\d+))?节?)?$", RegexOptions.Compiled);
        private static readonly Regex InheritedVerseWithJieRegex = new Regex(@"^第?(\d+)(?:-(\d+))?节$", RegexOptions.Compiled);
        private static readonly Regex InheritedVerseRangeRegex = new Regex(@"^(\d+)-(\d+)$", RegexOptions.Compiled);
        private static readonly Regex NumberOnlyRegex = new Regex(@"^(\d+)$", RegexOptions.Compiled);

        public ParsedReference ParseSelection(string rawSelection)
        {
            string normalized = NormalizeSelection(rawSelection);
            if (normalized.Length == 0)
            {
                throw new FormatException("没有选中文字");
            }

            string[] rawChunks = normalized.Split(',');
            List<PassageReference> passages = new List<PassageReference>();
            BibleBook currentBook = null;
            int? currentChapter = null;

            foreach (string rawChunk in rawChunks)
            {
                string chunk = rawChunk.Trim();
                if (chunk.Length == 0)
                {
                    continue;
                }

                ParsedChunk parsed = ParseChunk(chunk, currentBook, currentChapter);
                passages.Add(parsed.Passage);
                currentBook = parsed.Passage.Book;
                currentChapter = parsed.ContextChapter;
            }

            if (passages.Count == 0)
            {
                throw new FormatException("未识别到经文引用");
            }

            return new ParsedReference(passages);
        }

        public VerseReference Parse(string rawSelection)
        {
            ParsedReference parsed = ParseSelection(rawSelection);
            if (parsed.Passages.Count != 1)
            {
                throw new FormatException("未识别到经文引用");
            }

            PassageReference passage = parsed.Passages[0];
            if (passage.StartChapter != passage.EndChapter || !passage.StartVerse.HasValue || !passage.EndVerse.HasValue)
            {
                throw new FormatException("未识别到经文引用");
            }

            return new VerseReference(passage.Book, passage.StartChapter, passage.StartVerse.Value, passage.EndVerse.Value);
        }

        private static string NormalizeSelection(string raw)
        {
            string text = (raw ?? string.Empty).Trim()
                .Replace('\n', ' ')
                .Replace('\t', ' ')
                .Replace('\u3000', ' ')
                .Replace("：", ":")
                .Replace("﹕", ":")
                .Replace("“", string.Empty)
                .Replace("”", string.Empty)
                .Replace("\"", string.Empty)
                .Replace("'", string.Empty);

            text = ConvertFullWidthDigits(text);
            text = NormalizeRanges(text);
            text = NormalizeSeparators(text);
            while (text.Contains("  "))
            {
                text = text.Replace("  ", " ");
            }
            return text.Trim();
        }

        private static string NormalizeRanges(string text)
        {
            string[] tokens = { "……", "...", "——", "--", "－", "–", "—", "―", "﹣", "～", "~", "^", "到", "至" };
            foreach (string token in tokens)
            {
                text = text.Replace(token, "-");
            }

            return Regex.Replace(text, @"(?i)(\d)\s*to\s*(\d)", "$1-$2");
        }

        private static string NormalizeSeparators(string text)
        {
            string[] tokens = { "，", "、", "；", ";", "｜", "|", "\\" };
            foreach (string token in tokens)
            {
                text = text.Replace(token, ",");
            }
            return text;
        }

        private static string ConvertFullWidthDigits(string text)
        {
            StringBuilder builder = new StringBuilder(text.Length);
            foreach (char ch in text)
            {
                if (ch >= '０' && ch <= '９')
                {
                    builder.Append((char)('0' + (ch - '０')));
                }
                else
                {
                    builder.Append(ch);
                }
            }
            return builder.ToString();
        }

        private static ParsedChunk ParseChunk(string rawChunk, BibleBook currentBook, int? currentChapter)
        {
            string compact = rawChunk.Trim().ToLowerInvariant().Replace(" ", string.Empty);
            if (compact.Length == 0)
            {
                throw new FormatException("未识别到经文引用");
            }

            BibleBook book = currentBook;
            string body = compact;
            BibleBook matchedBook;
            string remaining;
            if (BibleBookCatalog.FindAtStart(compact, out matchedBook, out remaining))
            {
                book = matchedBook;
                body = remaining;
            }

            if (book == null)
            {
                throw new FormatException("未识别书卷：" + rawChunk);
            }

            PassageReference passage = ParseChapterStyle(body, book, currentChapter)
                ?? ParseColonStyle(body, book)
                ?? ParseInheritedVerseStyle(body, book, currentChapter);

            if (passage == null)
            {
                throw new FormatException("未识别到经文引用");
            }

            return new ParsedChunk(passage, passage.EndChapter);
        }

        private static PassageReference ParseChapterStyle(string body, BibleBook book, int? currentChapter)
        {
            if (!body.Contains("章") && !body.Contains("节"))
            {
                return null;
            }

            Match match = WholeChapterRegex.Match(body);
            if (match.Success)
            {
                int chapter = int.Parse(match.Groups[1].Value);
                int? startVerse = match.Groups[2].Success ? (int?)int.Parse(match.Groups[2].Value) : null;
                int? endVerse = match.Groups[3].Success ? (int?)int.Parse(match.Groups[3].Value) : startVerse;
                if (startVerse.HasValue && endVerse.HasValue && startVerse.Value > endVerse.Value)
                {
                    throw new FormatException("范围顺序不正确");
                }
                return new PassageReference(book, chapter, startVerse, chapter, endVerse);
            }

            match = InheritedVerseWithJieRegex.Match(body);
            if (match.Success && currentChapter.HasValue)
            {
                int startVerse = int.Parse(match.Groups[1].Value);
                int endVerse = match.Groups[2].Success ? int.Parse(match.Groups[2].Value) : startVerse;
                if (startVerse > endVerse)
                {
                    throw new FormatException("范围顺序不正确");
                }
                return new PassageReference(book, currentChapter.Value, startVerse, currentChapter.Value, endVerse);
            }

            return null;
        }

        private static PassageReference ParseColonStyle(string body, BibleBook book)
        {
            Match match = CrossChapterRegex.Match(body);
            if (match.Success)
            {
                int startChapter = int.Parse(match.Groups[1].Value);
                int startVerse = int.Parse(match.Groups[2].Value);
                int endChapter = int.Parse(match.Groups[3].Value);
                int endVerse = int.Parse(match.Groups[4].Value);
                if (startChapter > endChapter || (startChapter == endChapter && startVerse > endVerse))
                {
                    throw new FormatException("范围顺序不正确");
                }
                return new PassageReference(book, startChapter, startVerse, endChapter, endVerse);
            }

            match = SameChapterRangeRegex.Match(body);
            if (match.Success)
            {
                int chapter = int.Parse(match.Groups[1].Value);
                int startVerse = int.Parse(match.Groups[2].Value);
                int endVerse = int.Parse(match.Groups[3].Value);
                if (startVerse > endVerse)
                {
                    throw new FormatException("范围顺序不正确");
                }
                return new PassageReference(book, chapter, startVerse, chapter, endVerse);
            }

            match = SingleVerseRegex.Match(body);
            if (match.Success)
            {
                int chapter = int.Parse(match.Groups[1].Value);
                int verse = int.Parse(match.Groups[2].Value);
                return new PassageReference(book, chapter, verse, chapter, verse);
            }

            return null;
        }

        private static PassageReference ParseInheritedVerseStyle(string body, BibleBook book, int? currentChapter)
        {
            Match match = InheritedVerseRangeRegex.Match(body);
            if (match.Success && currentChapter.HasValue)
            {
                int startVerse = int.Parse(match.Groups[1].Value);
                int endVerse = int.Parse(match.Groups[2].Value);
                if (startVerse > endVerse)
                {
                    throw new FormatException("范围顺序不正确");
                }
                return new PassageReference(book, currentChapter.Value, startVerse, currentChapter.Value, endVerse);
            }

            match = NumberOnlyRegex.Match(body);
            if (match.Success)
            {
                int number = int.Parse(match.Groups[1].Value);
                if (currentChapter.HasValue)
                {
                    return new PassageReference(book, currentChapter.Value, number, currentChapter.Value, number);
                }
                return new PassageReference(book, number, null, number, null);
            }

            return null;
        }

        private sealed class ParsedChunk
        {
            public ParsedChunk(PassageReference passage, int contextChapter)
            {
                Passage = passage;
                ContextChapter = contextChapter;
            }

            public PassageReference Passage { get; private set; }
            public int ContextChapter { get; private set; }
        }
    }
}
